enumerables.rb

```diff
@@ -0,0 +1,142 @@
+# rubocop: disable Metrics/ModuleLength, Metrics/MethodLength, Style/CaseEquality, Style/IfInsideElse
+
+module Enumerable
+  def my_each(var = nil)
+    return to_enum unless block_given?
+
+    n = if var.nil?
+          0
+        else
+          var
+        end
+    while n <= size - 1
+      yield(to_a[n])
+      n += 1
+    end
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    n = 0
+    while n <= size - 1
+      yield(to_a[n], n)
+      n += 1
+    end
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    result = []
+    my_each do |n|
+      result.push(n) if yield(n)
+    end
+    result
+  end
+
+  def my_all?(var = nil)
+    result = true
+    my_each do |n|
+      if block_given?
+        result = false unless yield(n)
+      elsif var.nil?
+        result = false unless n
+      else
+        result = false unless var === n
+      end
+    end
+    result
+  end
+
+  def my_any?(var = nil)
+    result = false
+    my_each do |n|
+      if block_given?
+        return result = true if yield(n)
+      elsif var.nil?
+        return result = true if n
+      else
+        return result = true if var === n
+      end
+    end
+    result
+  end
+
+  def my_none?(var = nil)
+    result = true
+    my_each do |n|
+      if block_given?
+        return result = false if yield(n)
+      elsif var.nil?
+        return result = false if n
+      else
+        return result = false if var === n
+      end
+    end
+    result
+  end
+
+  def my_count(var = nil)
+    result = []
+    if block_given?
+      my_each do |n|
+        result.push(n) if yield(n)
+      end
+    elsif var.nil?
+      return length
+    else
+      my_each do |n|
+        result.push(n) if n == var
+      end
+    end
+    result.length
+  end
+
+  def my_map(var = nil)
+    result = []
+    if block_given? && var.nil?
+      my_each do |n|
+        result.push(yield(n))
+      end
+    else
+      my_each do |n|
+        result.push(var.call(n))
+      end
+    end
+    result
+  end
```

Comment:
* [ ] my_map returns an Enumerator if no block is given
   ```ruby
    array = [1,5,8,7,9,5,5,55,5]
    p array.my_map #error
    # NoMethodError:
       undefined method `call' for nil:NilClass
     # ./enumerables.rb:104:in `block in my_map'
     # ./enumerables.rb:13:in `my_each'
     # ./enumerables.rb:103:in `my_map'
  ```

Reply:
Ok, just changed that.